### ISSUE TITLE
`OnPostDispatch` hook

### ISSUE THREAD

**Issue by kianenigma** [paritytech/substrate#12047]
This came up during a FRAME call. Imagine: 

```rust
#[pallet::hooks]
impl<T: Config> Hooks for Pallet<T> {
  fn on_post_dispatch() -> DispatchResult { 
    // ANY dispatch is valid IFF this function also returns `Ok()`. 
    // You better be really careful about implementing this! 
  }
}
```

Combined with transactional by default, we can then greatly simplify the balances pallet and do a existential deposit check after every transaction.

One question would be why and how would you use this if your pallet is not transactional? should it even be allowed?

**Comment by shawntabrizi** [comment]
> One question would be why and how would you use this if your pallet is not transactional? should it even be allowed?

FRAME cannot support both transactional and non transactional. We have to be entirely one or the other. In this case, we have chosen to make EVERYTHING transactional, and thus this kind of thing should now be allowable. (whereas it wasn't before)

We should not consider the case where things are not transactional, since this is just simply not supported.

**Comment by bkchr** [comment]
Hmmm. Can someone give a concrete example on what this should be used for? 

These `Hooks` are also currently implemented for the `Pallet ` and not the `Call`. So, the pseudo code wouldn't really work :P We could pass the `Call` there as an argument. Or we support some kind of `post_dispatch` attribute. However, I'm still wondering what would be the purpose, as you can do these post dispatch checks at the end of the call?

**Comment by kianenigma** [comment]
> Hmmm. Can someone give a concrete example on what this should be used for? 


> greatly simplify the balances pallet and do a existential deposit check after every transaction.

The pseudo code is there to demonstrate a dispatch is valid of _both its dispatch and post_dispatch_ are 'Ok', thus 'x.and(y)'.

**Comment by bkchr** [comment]
> Of course you can just do this in tue dispatch, but wouldn't you want to factor it out and not need to check it everywhere?

By check everywhere you mean at each possible return of your dispatchable?

**Comment by kianenigma** [comment]
> By check everywhere you mean at each possible return of your dispatchable?

Yes.

**Comment by ggwpez** [comment]
I think this would be more useful on a per-extrinsic level, similar to how we have the `verify` block in benchmarks.  
Some kind of post-conditions check is in general a good idea IMHO. Especially for testing because we can stuff it with `debug_asserts`.

**Comment by bkchr** [comment]
For per-extrinsic level we already have `SignedExtension` that have a `post_dispatch`.

**Comment by ruseinov** [comment]
I'm not 100% sure the following belongs in this issue, however it's not dissimilar:

As discussed in the FRAME channel and during the last FRAME Steering call - it would be nice to introduce something that would be executed before each extrinsic of the given pallet:
```rust
// not sure if this should live in `hooks`, but why not
#[pallet::hooks]
impl<T: Config> Hooks for Pallet<T> {
    on_pre_call(call: Call) -> DispatchResult {
        
    }
}
```

One good usage example is the need to disable certain (or all) extrinsics while the pallet is running a multi-block migration.
Something like this:
```rust
#[pallet::hooks]
impl<T: Config> Hooks for Pallet<T> {
    on_pre_call(call: Call) -> DispatchResult {
      let migration_is_running = read_temporary_storage_item();
      if migration_is_running {
           return Err(TemporaryDisabled)
      }
      return Ok(())
    }
}
```

Off the top of my head the way this could work is `#[pallet::call]` macro would go and add the `on_pre_call` to each extrinsic.

As discussed in the call it could also be done by manually introducing an `if` condition to every extrinsic. But this is:
1. Error-prone, especially when there are a lot of extrinsics to cover.
2. Not very DRY
3. Harder to test, as each extrinsic then would need to be tested separately.

**Comment by bkchr** [comment]
When there is a multi block migration is running, we should disable all pallets and only accept inherents. We should not only disable some pallets.

**Comment by ruseinov** [comment]
> When there is a multi block migration is running, we should disable all pallets and only accept inherents. We should not only disable some pallets.

Yep, I remember this discussion in another issue. Would be great to get this going.

**Comment by ruseinov** [comment]
> When there is a multi block migration is running, we should disable all pallets and only accept inherents. We should not only disable some pallets.

Yep, I remember this discussion in another issue. Would be great to get this going.

**Comment by kianenigma** [comment]
> er-extrinsic level we already have `SignedExtension` that have a `post_disp`

- To recap this, global pre/post hooks must be set via signed extension, at least for now.
- This issue is more about a pre/post that is local to the dispatchables of this pallet.
